mgr: re-enable mds `scrub status` info in ceph status

vshankar opened the pull request:
@tchaikov @batrick 

MDS does not register with the manager for service updates and is not a service daemon. This pr adjusts `task status` update logic for that. One thing to note here is that the MDS (in the form of `<service>.<name>`)  gets added to the service map though it's not a service daemon. Would that cause any side effects?

@batrick -- right now I haven't changed the way MDS constantly sends scrub updates even when idle ("idle" status). I'll get to that once the approach here seems ok.

batrick commented:
> @tchaikov @batrick
> 
> MDS does not register with the manager for service updates and is not a service daemon. This pr adjusts `task status` update logic for that. One thing to note here is that the MDS (in the form of `<service>.<name>`) gets added to the service map though it's not a service daemon. Would that cause any side effects?

It might affect `ceph versions` output. Have you checked? That was one of the big issues with the MDS registering itself before.

> @batrick -- right now I haven't changed the way MDS constantly sends scrub updates even when idle ("idle" status). I'll get to that once the approach here seems ok.

You mean so that the `ceph status` output does not show an idle scrub all the time? Yes, I think we definitely need to fix that but it doesn't need to be in this patch.

batrick commented:
@vshankar please run this through teuthology to confirm it fixes the test failures.

vshankar commented:
> > @tchaikov @batrick
> > MDS does not register with the manager for service updates and is not a service daemon. This pr adjusts `task status` update logic for that. One thing to note here is that the MDS (in the form of `<service>.<name>`) gets added to the service map though it's not a service daemon. Would that cause any side effects?
> 
> It might affect `ceph versions` output. Have you checked? That was one of the big issues with the MDS registering itself before.

I went through the related tracker tickets from the earlier issue -- one of them was related to mds metadata. `ceph versions` (additionally) dumps services from service map here: https://github.com/ceph/ceph/blob/master/src/mon/Monitor.cc#L3799

That will dump an (additional) MDS entry with an unknown version. Not sure if filtering it would be good enough, @tchaikov any thoughts?

> 
> > @batrick -- right now I haven't changed the way MDS constantly sends scrub updates even when idle ("idle" status). I'll get to that once the approach here seems ok.
> 
> You mean so that the `ceph status` output does not show an idle scrub all the time? Yes, I think we definitely need to fix that but it doesn't need to be in this patch.

sure.

vshankar commented:
@tchaikov ping

vshankar commented:
http://pulpito.ceph.com/vshankar-2020-01-23_14:00:03-rados-wip-vshankar-mds-task-status-distro-basic-smithi/

vshankar commented:
> http://pulpito.ceph.com/vshankar-2020-01-23_14:00:03-rados-wip-vshankar-mds-task-status-distro-basic-smithi/

@batrick @tchaikov -- run looks ok to me -- there's one failure, but looks unrelated (I'll take another look tomorrow). I added a commit to filter normal ceph entity types when dumping service metadata.

Note that the above rados suite run is with "--limit 100". @tchaikov -- please take a look at this PR and include it in your next run if it looks ok?

vshankar commented:
> ```
> 2020-02-06T10:59:25.403 INFO:teuthology.orchestra.run.smithi136:> sudo adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage timeout 120 ceph --cluster ceph --log-early fs status
> ...
> 2020-02-06T10:59:25.740 INFO:teuthology.orchestra.run.smithi136.stderr:Error EINVAL: AttributeError: 'NoneType' object has no attribute 'get'
> 2020-02-06T10:59:25.749 DEBUG:teuthology.orchestra.run:got remote process result: 22
> 2020-02-06T10:59:25.750 INFO:tasks.cephfs_test_runner:test_fs_status (tasks.cephfs.test_admin.TestAdminCommands) ... ERROR
> ```
> 
> 
> /ceph/teuthology-archive/pdonnell-2020-02-06_01:51:32-fs-wip-pdonnell-testing-20200205.204218-distro-basic-smithi/4736844/teuthology.log
> 
> 
> `ceph fs status` is broken by this PR

will take a look

vshankar commented:
> `ceph fs status` is broken by this PR

This is due to the side effect of adding mds daemon entries in service map. I need to *carefully* audit operations performed using service map in manager and monitor -- some usages are not obvious causing failures as seen here.

I'll push and update when I'm done auditing and fixing.

vshankar commented:
fixed and updated

@batrick -- run link here (w/ the fix): http://pulpito.ceph.com/vshankar-2020-02-09_09:15:31-fs-wip-mds-scrub-task-status-distro-basic-smithi/

please take a look.

batrick commented:
jenkins test make check

batrick commented:
retest this please

batrick commented:
You meant to fix this one too right?

````
2020-02-15T18:23:44.118 INFO:tasks.cephfs_test_runner:======================================================================
2020-02-15T18:23:44.118 INFO:tasks.cephfs_test_runner:FAIL: test_scrub_pause_and_resume (tasks.cephfs.test_scrub_checks.TestScrubControls)
2020-02-15T18:23:44.118 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-02-15T18:23:44.118 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-02-15T18:23:44.118 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20200215.033325/qa/tasks/cephfs/test_scrub_checks.py", line 81, in test_scrub_pause_and_resume
2020-02-15T18:23:44.118 INFO:tasks.cephfs_test_runner:    self._check_task_status("paused")
2020-02-15T18:23:44.118 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-pdonnell-testing-20200215.033325/qa/tasks/cephfs/test_scrub_checks.py", line 35, in _check_task_status
2020-02-15T18:23:44.119 INFO:tasks.cephfs_test_runner:    self.assertTrue(task_status['a'].startswith(expected_status))
2020-02-15T18:23:44.119 INFO:tasks.cephfs_test_runner:AssertionError: False is not true
2020-02-15T18:23:44.119 INFO:tasks.cephfs_test_runner:
````

/ceph/teuthology-archive/pdonnell-2020-02-15_16:51:06-fs-wip-pdonnell-testing-20200215.033325-distro-basic-smithi/4767466/teuthology.log

vshankar commented:
> You meant to fix this one too right?

Shouldn't this test be running with a single active MDS since `MDSS_REQUIRED = 1`? Task status is seeing two active MDSs:

`2020-02-15T18:23:13.677 INFO:tasks.cephfs.filesystem.ceph_manager:{u'a': u'idle', u'b': u'paused'}
`
The test does not change `max_mds` too..

vshankar commented:
> Shouldn't this test be running with a single active MDS since `MDSS_REQUIRED = 1`? Task status is seeing two active MDSs:
> 
> `2020-02-15T18:23:13.677 INFO:tasks.cephfs.filesystem.ceph_manager:{u'a': u'idle', u'b': u'paused'} `
> The test does not change `max_mds` too..

spoke too soon: the other mds is a standby and it should not be sending scrub status updated.

vshankar commented:
@batrick it's not all that bad -- mds.a and mds.b are switching from active -> standby -> active.

mds.a:

```
2020-02-15T18:17:43.045+0000 7f159ee31700  1 mds.a Monitors have assigned me to become a standby.
2020-02-15T18:18:19.685+0000 7fe0d832d700  1 mds.a Monitors have assigned me to become a standby.
2020-02-15T18:18:22.043+0000 7fe0d832d700  1 mds.0.10 handle_mds_map state change up:boot --> up:creating
2020-02-15T18:18:23.047+0000 7fe0d832d700  1 mds.0.10 handle_mds_map state change up:creating --> up:active
2020-02-15T18:19:47.247+0000 7f98d0fca700  1 mds.a Monitors have assigned me to become a standby.
2020-02-15T18:19:50.615+0000 7f98d0fca700  1 mds.0.18 handle_mds_map state change up:boot --> up:creating
2020-02-15T18:19:51.624+0000 7f98d0fca700  1 mds.0.18 handle_mds_map state change up:creating --> up:active
2020-02-15T18:20:06.078+0000 7fad2707b700  1 mds.a Monitors have assigned me to become a standby.
2020-02-15T18:21:00.471+0000 7fa4dea96700  1 mds.a Monitors have assigned me to become a standby.
2020-02-15T18:21:03.837+0000 7fa4dea96700  1 mds.0.29 handle_mds_map state change up:boot --> up:creating
2020-02-15T18:21:04.840+0000 7fa4dea96700  1 mds.0.29 handle_mds_map state change up:creating --> up:active
2020-02-15T18:22:22.051+0000 7fd43cf72700  1 mds.a Monitors have assigned me to become a standby.
2020-02-15T18:22:28.470+0000 7f5e4d388700  1 mds.a Monitors have assigned me to become a standby.

```

mds.b:

```
2020-02-15T18:17:43.045+0000 7fcab041e700  1 mds.0.4 handle_mds_map state change up:boot --> up:creating
2020-02-15T18:17:44.051+0000 7fcab041e700  1 mds.0.4 handle_mds_map state change up:creating --> up:active
 -2626> 2020-02-15T18:17:43.045+0000 7fcab041e700  1 mds.0.4 handle_mds_map state change up:boot --> up:creating
 -1831> 2020-02-15T18:17:44.051+0000 7fcab041e700  1 mds.0.4 handle_mds_map state change up:creating --> up:active
2020-02-15T18:18:13.080+0000 7fb4bb8e8700  1 mds.b Monitors have assigned me to become a standby.
2020-02-15T18:18:16.127+0000 7fe1988d3700  1 mds.b Monitors have assigned me to become a standby.
2020-02-15T18:19:47.247+0000 7f2cff10b700  1 mds.b Monitors have assigned me to become a standby.
2020-02-15T18:20:06.076+0000 7ffae8a3f700  1 mds.b Monitors have assigned me to become a standby.
2020-02-15T18:20:06.082+0000 7ffae8a3f700  1 mds.0.22 handle_mds_map state change up:boot --> up:replay
2020-02-15T18:20:07.087+0000 7ffae8a3f700  1 mds.0.22 handle_mds_map state change up:replay --> up:reconnect
2020-02-15T18:20:08.088+0000 7ffae8a3f700  1 mds.0.22 handle_mds_map state change up:reconnect --> up:rejoin
2020-02-15T18:20:09.089+0000 7ffae8a3f700  1 mds.0.22 handle_mds_map state change up:rejoin --> up:active
 -4039> 2020-02-15T18:20:06.076+0000 7ffae8a3f700  1 mds.b Monitors have assigned me to become a standby.
 -3999> 2020-02-15T18:20:06.082+0000 7ffae8a3f700  1 mds.0.22 handle_mds_map state change up:boot --> up:replay
 -3741> 2020-02-15T18:20:07.087+0000 7ffae8a3f700  1 mds.0.22 handle_mds_map state change up:replay --> up:reconnect
 -3692> 2020-02-15T18:20:08.088+0000 7ffae8a3f700  1 mds.0.22 handle_mds_map state change up:reconnect --> up:rejoin
 -3625> 2020-02-15T18:20:09.089+0000 7ffae8a3f700  1 mds.0.22 handle_mds_map state change up:rejoin --> up:active
2020-02-15T18:20:57.461+0000 7f4cb3b4e700  1 mds.b Monitors have assigned me to become a standby.
2020-02-15T18:21:00.471+0000 7f333b594700  1 mds.b Monitors have assigned me to become a standby.
2020-02-15T18:22:25.093+0000 7f58e1c39700  1 mds.b Monitors have assigned me to become a standby.
2020-02-15T18:22:28.470+0000 7f58e1c39700  1 mds.0.36 handle_mds_map state change up:boot --> up:creating
2020-02-15T18:22:29.471+0000 7f58e1c39700  1 mds.0.36 handle_mds_map state change up:creating --> up:active
```

`task_status` would prune service map entries that have not responded for more than `mgr_service_beacon_grace` seconds. Until then `task_status` would (briefly) contain entries from both MDSs. The test case would need to query the currently active MDS.

vshankar commented:
@batrick fixed and update -- this should (hopefully) be the last of the failures...
